pkg/tracing/tracer.go

```diff
@@ -3,66 +3,68 @@ package tracing
 import (
 	"context"
 	"fmt"
-	"net/http"
-	"sync"
 
+	"github.com/justtrackio/gosoline/pkg/appctx"
 	"github.com/justtrackio/gosoline/pkg/cfg"
 	"github.com/justtrackio/gosoline/pkg/log"
 )
 
 //go:generate mockery --name Tracer
 type Tracer interface {
-	HttpHandler(h http.Handler) http.Handler
 	StartSpan(name string) (context.Context, Span)
 	StartSpanFromContext(ctx context.Context, name string) (context.Context, Span)
 	StartSubSpan(ctx context.Context, name string) (context.Context, Span)
 }
 
 type TracerSettings struct {
-	Provider                    string                `cfg:"provider" default:"xray" validate:"required"`
-	Enabled                     bool                  `cfg:"enabled" default:"false"`
-	AddressType                 string                `cfg:"addr_type" default:"local" validate:"required"`
-	AddressValue                string                `cfg:"add_value" default:""`
-	Sampling                    SamplingConfiguration `cfg:"sampling"`
-	StreamingMaxSubsegmentCount int                   `cfg:"streaming_max_subsegment_count" default:"20"`
+	Provider string `cfg:"provider"`
+	Enabled  bool   `cfg:"enabled" default:"false"`
 }
 
-var tracerContainer = struct {
-	sync.Mutex
-	instance Tracer
-}{}
+type appCtxKey string
 
-func ProvideTracer(config cfg.Config, logger log.Logger) (Tracer, error) {
-	tracerContainer.Lock()
-	defer tracerContainer.Unlock()
+func ProvideTracer(ctx context.Context, config cfg.Config, logger log.Logger) (Tracer, error) {
+	return appctx.Provide(ctx, appCtxKey("tracer"), func() (Tracer, error) {
+		return NewTracer(ctx, config, logger)
+	})
+}
+
+func ProvideInstrumentor(ctx context.Context, config cfg.Config, logger log.Logger) (Instrumentor, error) {
+	return appctx.Provide(ctx, appCtxKey("instrumentor"), func() (Instrumentor, error) {
+		return NewInstrumentor(ctx, config, logger)
+	})
+}
+
+func NewTracer(ctx context.Context, config cfg.Config, logger log.Logger) (Tracer, error) {
```

Comment:
Why is NewTracer and NewInstrumentor public when we have Provide* methods? Do we ever need more than one?